From 7a770be05f798f977be20add9ac35e2558d13bd6 Mon Sep 17 00:00:00 2001
From: Kiran Rudramuni <kchitrik@codeaurora.org>
Date: Fri, 20 Mar 2015 16:20:19 -0700
Subject: [PATCH] msm_shared: hdmi: check for DVI before configuring audio

Do not configure audio in case sink is DVI. Also, set Tx into
DVI mode.

Change-Id: I364f77b131231a14debcd133601d47315c45d646
---
 platform/msm_shared/mdss_hdmi.c | 46 +++++++++++++++++++++++++++++++++--------
 1 file changed, 37 insertions(+), 9 deletions(-)

diff --git a/platform/msm_shared/mdss_hdmi.c b/platform/msm_shared/mdss_hdmi.c
index e5a27ea..91997a6 100644
--- a/platform/msm_shared/mdss_hdmi.c
+++ b/platform/msm_shared/mdss_hdmi.c
@@ -611,18 +611,45 @@ bail_ldo_fail:
 	return ret;
 }
 
+static bool mdss_hdmi_is_dvi_mode(void)
+{
+	uint8_t len;
+	uint32_t ieee_tag;
+	const uint8_t *vsd = NULL;
+
+	vsd = hdmi_edid_find_block(DBC_START_OFFSET,
+		VENDOR_SPECIFIC_DATA_BLOCK, &len);
+
+	if (vsd == NULL || len == 0) {
+		dprintf(SPEW, "%s: Invalid VSDB\n", __func__);
+                return false;
+	}
+
+	ieee_tag = ((uint32_t) vsd[3] << 16) + ((uint32_t) vsd[2] << 8) +
+			(uint32_t) vsd[1];
+
+	if (ieee_tag == 0x0c03)
+		return false;
+	else
+		return true;
+}
+
 static void mdss_hdmi_set_mode(bool on)
 {
 	uint32_t val = 0;
 
 	if (on) {
-		val = 0x3;
-		writel(val, HDMI_CTRL);
-	} else {
-		writel(val, HDMI_CTRL);
+		/* tx on */
+		val = 0x1;
+
+		if (!mdss_hdmi_is_dvi_mode())
+			val |= 0x2;
 	}
+
+	writel(val, HDMI_CTRL);
 }
 
+
 static int mdss_hdmi_read_edid(void)
 {
 	uint8_t ndx;
@@ -1037,21 +1064,22 @@ void mdss_hdmi_avi_info_frame(void)
 
 int mdss_hdmi_init(void)
 {
+	bool is_dvi_mode = mdss_hdmi_is_dvi_mode();
+
 	mdss_hdmi_set_mode(false);
 
 	hdmi_phy_init();
 
-	/* Enable USEC REF timer */
-	writel(0x0001001B, HDMI_USEC_REFTIMER);
-
 	/* Audio settings */
-	mdss_hdmi_audio_playback();
+	if (!is_dvi_mode)
+		mdss_hdmi_audio_playback();
 
 	/* Video settings */
 	mdss_hdmi_video_setup();
 
 	/* AVI info settings */
-	mdss_hdmi_avi_info_frame();
+	if (!is_dvi_mode)
+		mdss_hdmi_avi_info_frame();
 
 	/* Enable HDMI */
 	mdss_hdmi_set_mode(true);
-- 
1.8.2.1

